Refactor code to create a component for the `in the last X days` part of some dynamic segments

rodrigoprimo opened the pull request:
## Description

As a first step to implementing changes to the `in the last X days` part of some dynamic segments (see [MAILPOET-4991]), in this PR, I refactored the code so that all the applicable dynamic segments use the same component for `in the last X days` part.

The affected dynamic segments are:

- `# of opens`
- `# of machine opens`
- `average order value`
- `# of orders`
- `single order value`
- `total spent`

## Code review notes

_N/A_

## QA notes

This PR should not change any user-facing behavior. It is probably a good idea to check that the functionality not covered by automated tests of the dynamic segments mentioned above remains the same.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4991]

## After-merge notes

_N/A_

## Tasks

- [X] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [X] I added sufficient test coverage
- [X] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-4991]: https://mailpoet.atlassian.net/browse/MAILPOET-4991?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-4991]: https://mailpoet.atlassian.net/browse/MAILPOET-4991?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Extract the days period selector into its own component

This commit extracts the days period selector from the
EmailOpensAbsoluteCountFields component into its own component called
DaysPeriodField. In the next commit, I will change other field
components to use DaysPeriodField its of duplicating the code to
generate the days period selector.

[MAILPOET-4991]

rodrigoprimo committed:
Refactor average spent and # of orders to use DaysPeriodField

This commit refactors the React components AverageSpentFields and
NumberOfOrdersFields to use the new DaysPeriodField component instead of
a its own code to generate the days period selector.

[MAILPOET-4991]

rodrigoprimo committed:
Refactor single order value and total spent to use DaysPeriodField

This commit refactors the React components SingleOrderValueFields and
TotalSpentFields to use the new DaysPeriodField component instead of
its own code to generate the days period selector.

[MAILPOET-4991]

johnolek reviewed (changes requested):
This looks good to me for filters going forward, but I think it would break existing filters that have data saved under the old keys like `single_order_value_days`. I think we'll either need to include a migration to update existing filters or have the code still check for the existence of the old key and use it if there isn't a value present for the `days` key.

rodrigoprimo commented:
Good point! I completely forgot to consider existing filters. I will work on it. Thanks.

rodrigoprimo committed:
Add migration to change the update name of parameters of a few filters

In a previous commit, the name of the parameter of a few filters was
changed to `days`. This commit adds a migration to change the name of
parameter of existing filters when needed. The following parameters
should be renamed to days:

- number_of_orders_days
- total_spent_days
- single_order_value_days
- average_spent_days

I opted to leave the original parameter instead of deleting it, just to
be safe in case a given site needs to rollback to a previous version.
Once a change is made to a filter by the user, the old parameter will be
deleted.

[MAILPOET-4991]

rodrigoprimo commented:
@johnolek, I ended up deciding to implement a migration to update the name of the parameters instead of maintaining the old keys working. I decided to do this when I saw that we had already migrations in the past that changed other parts of the dynamic segments (for example, https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/lib/Migrations/Migration_20221028_105818.php#L804 and https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/lib/Migrations/Migration_20221028_105818.php#L845).

I opted to preserve the old parameter and copy its value to the new one. This seemed a bit safer as it allows users to rollback to previous versions of MailPoet. The old parameter will be deleted if the user updates the segment.

Let me know what you think of this approach. This PR is ready for another review.

I had to force push to this branch to fix a conflict with trunk.

johnolek commented on mailpoet/lib/Migrations/Migration_20230712_180341.php:
```diff
@@ -0,0 +1,53 @@
+
+      if (isset($data['number_of_orders_days'])) {
+        $days = $data['number_of_orders_days'];
+      } else if (isset($data['total_spent_days'])) {
```
Just curious, is there a reason you're preferring `else if` to `elseif` here? It looks like the [WP standard](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#use-elseif-not-else-if) is to use `elseif`, but of course we aren't following WP standards in general.

rodrigoprimo replied:
No reason other than muscle memory. Despite the WP coding standard, I'm more used to `else if` than `elseif`. Happy to change it if you prefer. I'm not sure if it is worth it or not.

johnolek replied:
No preference here, just curious!

johnolek commented on mailpoet/tests/integration/Migrations/Migration_20230712_180341_Test.php:
```diff
@@ -0,0 +1,95 @@
+<?php declare(strict_types = 1);
+
+namespace integration\Migrations;
```
As far as I can tell it doesn't matter, but the one other migration test we have uses the `MailPoet\Migrations` namespace. Is there a reason you went with this instead?

rodrigoprimo replied:
My best guess is that this was autogenerated by PHPStorm and I didn't notice. I don't I added this. I just pushed a commit changing the namespace.

johnolek reviewed (dismissed):
Thank you for the explanation and for digging up those examples! I really like this approach, especially allowing for rollbacks.

I have a couple of very minor questions, but nothing here needs changing that I can see. Nice work!

Feel free to pass along to QA if you don't plan on making any tweaks.

rodrigoprimo committed:
Update integration test namespace

[MAILPOET-4991]

Aschepikov reviewed (approved):
GJ!
